homeassistant/components/matter/__init__.py

```diff
@@ -0,0 +1,234 @@
+"""The Matter integration."""
+from __future__ import annotations
+
+import asyncio
+from pathlib import Path
+from typing import cast
+
+import async_timeout
+from matter_server.client.exceptions import CannotConnect, FailedCommand
+from matter_server.client.matter import Matter
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, ServiceCall, callback
+from homeassistant.exceptions import ConfigEntryNotReady, HomeAssistantError
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.service import async_register_admin_service
+
+from .adapter import MatterAdapter, get_matter_store
+from .const import DOMAIN
+from .device_platform import DEVICE_PLATFORM
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Matter from a config entry."""
+    matter = Matter(MatterAdapter(hass, entry))
+    try:
+        await matter.connect()
+    except CannotConnect as err:
+        raise ConfigEntryNotReady("Failed to connect to matter server") from err
+
+    except Exception as err:
+        matter.adapter.logger.exception("Failed to connect to matter server")
+        raise ConfigEntryNotReady(
+            "Unknown error connecting to the Matter server"
+        ) from err
+
+    async def on_hass_stop(event: Event) -> None:
+        """Handle incoming stop event from Home Assistant."""
+        await matter.disconnect()
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
+    )
+
+    matter.listen()
+    try:
+        async with async_timeout.timeout(30):
+            await matter.driver_ready.wait()
+    except asyncio.TimeoutError as err:
+        raise ConfigEntryNotReady("Matter driver not ready") from err
+
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+        _async_init_services(hass)
+
+    hass.data[DOMAIN][entry.entry_id] = matter
+
+    await hass.config_entries.async_forward_entry_setups(entry, DEVICE_PLATFORM)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, DEVICE_PLATFORM)
+
+    if unload_ok:
+        matter: Matter = hass.data[DOMAIN].pop(entry.entry_id)
+        await matter.disconnect()
+
+    return unload_ok
+
+
+async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Config entry is being removed."""
+    # Remove storage file.
+    storage_path = get_matter_store(hass, entry).path
```

Comment:
If we decide to move all data to the add-on, merging this will have the side effect that we need to migrate it too.

Reply:
We don't need to migrate anything until we release this PR.